### ISSUE TITLE
Edited error queue message lost due to serialization exception

### ISSUE THREAD

**Issue by cjberg** [rebus-org/Rebus#459]
I'm attempting to learn Rebus so I’ve created a simple test project that makes use of MSMQ for the transport. In one of the tests I let the handler throw an exception _unless_ the message contains the custom header "open-sesame". I didn’t set the header on purpose when the message was sent, thereby effectively having the message end up on the error queue, _as expected_.

Just for shits and giggles, I then tried moving the message back to the input queue, verifying that Rebus would retry handling the message and eventually return it to the error queue. Of course it did.

This is where it gets interesting. I attempted to edit the error queue message’s headers using Cogin’s QueueExplorer, adding my custom header to the already existing JSON header collection stored in MSMQ’s extension property. I saved the message and moved it back to the input queue, hoping to see the handler accept it this time around.

But for some reason I managed to break the header data when I edited it. Rebus thus threw an exception (included below) when it couldn’t deserialize the message. That’s cool, I screwed up. What’s less cool is that **the message got lost** in some digital black hole instead of being returned to the error queue. I didn’t expect that at all.

Is this a bug or something one simply needs to be aware of?

```
Rebus.Workers.ThreadBased.ThreadWorker WARN (Rebus 1 worker 1): An error occurred when attempting to receive transport message: System.Runtime.Serialization.SerializationException: Could not deserialize MSMQ extension for message with physical message ID 01667298-8e88-4195-8c67-1b94600f2b9f\42187 - expected valid JSON text, got '?{
  "rbs2-intent": "p2p",
  "rbs2-msg-id": "c6feb342-e706-45ad-8ff7-f0ee2937d4f1",
  "rbs2-senttime": "2016-05-24T11:36:53.81314+02:00",
  "rbs2-corr-id": "c6feb342-e706-45ad-8ff7-f0ee2937d4f1",
  "rbs2-corr-seq": "0",
  "rbs2-msg-type": "RebusTest.Messages.TestMessage, RebusTest.Messages",
  "rbs2-content-type": "application/json;charset=utf-8",
  "rbs2-error-details": "...lots...",
  "rbs2-source-queue": "rebus-test@ENIAC",
  "open-sesame": "1"
}' ---> Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: ?. Path '', line 0, position 0.
   at Newtonsoft.Json.JsonTextReader.ParseValue()
   at Newtonsoft.Json.JsonTextReader.Read()
   at Newtonsoft.Json.JsonReader.ReadAndMoveToContent()
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.ReadForType(JsonReader reader, JsonContract contract, Boolean hasConverter)
   at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.Deserialize(JsonReader reader, Type objectType, Boolean checkAdditionalContent)
   at Newtonsoft.Json.JsonSerializer.DeserializeInternal(JsonReader reader, Type objectType)
   at Newtonsoft.Json.JsonConvert.DeserializeObject(String value, Type type, JsonSerializerSettings settings)
   at Newtonsoft.Json.JsonConvert.DeserializeObject[T](String value, JsonSerializerSettings settings)
   at Rebus.Transport.Msmq.MsmqTransport.ExtensionSerializer.Deserialize(Byte[] bytes, String msmqMessageId)
   --- End of inner exception stack trace ---
   at Rebus.Transport.Msmq.MsmqTransport.ExtensionSerializer.Deserialize(Byte[] bytes, String msmqMessageId)
   at Rebus.Transport.Msmq.MsmqTransport.<Receive>d__14.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Workers.ThreadBased.ThreadWorker.<TryReceiveTransportMessage>d__16.MoveNext()
```

**Comment by mookid8000** [comment]
That sounds weird.... this is something I will have to investigate. Thanks for reporting it 👍

**Comment by mookid8000** [comment]
whoa! it turned out that an invalid MSMQ message extension would result in an exception in a way that would result in still committing the `MessageQueueTransaction` _as if the message had been properly handled!_

I've fixed it - it will be out in 0.99.59 as soon as the build server has finished crunching

**Comment by mookid8000** [comment]
thank you so much for and excellent bug report 😄 👍

**Comment by cjberg** [comment]
**Thanks** for looking into this so quickly! It wasn't on NuGet yet, but I built Rebus myself this morning and tested it. The message won't disappear anymore, which is great, but I also noticed it isn't moved to the error queue. It turns into a poison message that is reprocessed over and over again, possibly preventing other messages from being consumed at all (especially on a single-threaded Rebus instance).

On the other hand, this might be such a rare form of error that moving the message manually is a sufficient solution. Not losing the message is what's most important.

By the way, I found out the reason it happened in the first place was that the tool I used to edit the message added a UTF BOM without asking. A little too "helpful".
